src/golang/lib/engine/aq_engine.go

```diff
@@ -974,25 +974,23 @@ func (eng *aqEngine) execute(
 				}
 
 				notificationCtxMsg := ""
-				if execState.Error != nil {
-					notificationCtxMsg = execState.Error.Message()
+				if execState.HasSystemError() {
+					if execState.Error != nil {
+						notificationCtxMsg = execState.Error.Message()
+					} else {
+						notificationCtxMsg = "This is a system error with no further context provided." + shared.TipCreateBugReport
+					}
 				}
 
 				notificationContent = &notificationContentStruct{
-					level:      mdl_shared.ErrorNotificationLevel,
-					contextMsg: notificationCtxMsg,
+					level:            mdl_shared.ErrorNotificationLevel,
+					systemErrContext: notificationCtxMsg,
```

Comment:
This naming confuses me since the conditional for this block is HasBlockingFailure, which isn't necessarily a system failure. So on user failures the context error here is just an empty string? Can you explain how user errors are surfaced again?

Reply:
let me add some comments to `notificationStructs`. The user errors are captured by `wfDag` object using `CheckWithErrors()` and `CheckWithWarning()` methods.

But there's indeed a miss here! We should capture user error from other operators as well by exposing some proper API from wfDag